README.md
# Win32::Shortcut::Readlink [![Build Status](https://secure.travis-ci.org/plicease/Win32-Shortcut-Readlink.png)](http://travis-ci.org/plicease/Win32-Shortcut-Readlink)

Make readlink work with shortcuts

# SYNOPSIS

    use Win32::Shortcut::Readlink;
    
    my $target = readlink "c:\\users\\foo\\Desktop\\Application.lnk";

# DESCRIPTION

This module overloads the perl built-in function [readlink](https://metacpan.org/pod/perlfunc#readlink)
so that it will treat shortcuts like pseudo symlinks on `cygwin` and `MSWin32`.
This module doesn't do anything on any other platform, so you are free to make
this a dependency, even if your module or script is going to run on non-Windows
platforms.

This module adjusts the behavior of readlink ONLY in the calling module, so
you shouldn't have to worry about breaking other modules that depend on the
more traditional behavior.

# FUNCTION

## readlink

    my $target = readlink EXPR
    my $target = readlink

Returns the value of a symbolic link or the target of the shortcut on Windows,
if either symbolic links are implemented or if shortcuts are.  If not, raises an 
exception.  If there is a system error, returns the undefined value and sets 
`$!` (errno). If `EXPR` is omitted, uses `$_`.

# CAVEATS

Does not handle Unicode.  Patches welcome.

Before Perl 5.16, `CORE` functions could not be aliased, and you will see warnings
on Perl 5.14 and earlier if you pass undef in as the argument to readlink, even if
you have warnings turned off.  The work around is to make sure that you never pass
undef to readlink on Perl 5.14 or earlier.

Perl 5.8.x is somewhat supported.  The use of implicit `$_` with readlink in
Perl 5.8.x is not supported and will throw an exception.  It is recommended that
you either upgrade to at least Perl 5.10 or pass an explicit argument of readlink
when using this module.

# SEE ALSO

- [Win32::Shortcut](https://metacpan.org/pod/Win32::Shortcut)
- [Win32::Unicode::Shortcut](https://metacpan.org/pod/Win32::Unicode::Shortcut)
- [Win32::Symlink](https://metacpan.org/pod/Win32::Symlink)
- [Win32::Hardlink](https://metacpan.org/pod/Win32::Hardlink)

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.03      2017-08-31 14:08:11 -0400
  - Migrate to EUMM

0.02      2014-05-10 17:24:06 -0400
  - include partial support for Perl 5.8

0.01      2014-05-08 17:18:18 -0400
  - initial version


xs/resolve.cpp
#if defined(_WIN32) || defined(__CYGWIN__)

#include <stdlib.h>
#include <math.h>
#include <windows.h>

#include <shlobj.h>
#include <shlguid.h>
#include <objbase.h>

#else
#include <stdlib.h>
#endif

#include <win32sr.h>

#if defined(_WIN32) || defined(__CYGWIN__)

const char *resolve(const char *link_name)
{
  HRESULT hres;
  IShellLink* ilink;
  const char *answer = NULL;
  static char target_path[MAX_PATH];

  hres = CoCreateInstance(CLSID_ShellLink, NULL, CLSCTX_INPROC_SERVER,
                            IID_IShellLink, (void **) &ilink);
  if (SUCCEEDED(hres))
  {
    IPersistFile* ifile;
    hres = ilink->QueryInterface(IID_IPersistFile, (void **) &ifile);
    if(SUCCEEDED(hres))
    {
      wchar_t w_link_name[MAX_PATH];
      /* TODO: handle proper unicode stuff properly */
      MultiByteToWideChar(CP_ACP, 0, link_name, -1, w_link_name, MAX_PATH);
      hres = ifile->Load(w_link_name, STGM_READ);
      if(SUCCEEDED(hres))
      {
        DWORD flags = 0;
        WIN32_FIND_DATA file;
        hres = ilink->GetPath((LPSTR) target_path, MAX_PATH, &file, flags);
        if(SUCCEEDED(hres))
          answer = target_path;
      }
      ifile->Release();
    }
    ilink->Release();
  }
  return target_path;
}

#else

const char *resolve(const char *link_name)
{
  return NULL;
}

#endif


lib/Win32/Shortcut/Readlink/Perl56.pm
package
  Win32::Shortcut::Readlink;

use strict;
use warnings;
use Carp qw( carp );
use constant _is_cygwin  => $^O eq 'cygwin';
use constant _is_mswin32 => $^O eq 'MSWin32';
use constant _is_windows => _is_cygwin || _is_mswin32;

sub readlink (;$)
{
  if(_is_windows)
  {
    if(@_ > 0)
    { CORE::readlink($_) }
    else
    { CORE::readlink($_[0]) }
  }

  my $arg = @_ > 0 ? $_[0] : $_;
  
  if(defined $arg && $arg =~ /\.lnk$/ && -r $arg)
  {
    my $target = _win32_resolve(_is_cygwin ? Cygwin::posix_to_win_path($arg) : $arg);
    return $target if defined $target;
  }

  if(_is_cygwin)
  {
    if(@_ > 0)
    { CORE::readlink($_) }
    else
    { CORE::readlink($_[0]) }
  }

  # else is MSWin32
  # emulate unix failues
  if(!defined $arg)
  {
    # TODO: only warn if warnings on in caller
    carp "Use of uninitialized value in readlink";
    $! = 22;
  }
  elsif(-e $arg)
  {
    $! = 22; # Invalid argument
  }
  else
  {
    $! = 2; # No such file or directory
  }
  
  return;
}

1;



lib/Win32/Shortcut/Readlink/Perl58.pm
package
  Win32::Shortcut::Readlink;

use strict;
use warnings;
use Carp qw( carp croak );
use constant _is_cygwin  => $^O eq 'cygwin';
use constant _is_mswin32 => $^O eq 'MSWin32';
use constant _is_windows => _is_cygwin || _is_mswin32;

sub readlink (;$)
{
  unless(_is_windows)
  {
    if(@_ > 0)
    { return CORE::readlink($_[0]) }
    else
    # If you need this feature, then please either upgrade to 5.10 or downgrade to 5.6
    { croak "implicit \$_ and Win32::Shortcut::Readlink are not interoperable on Perl 5.8.x" }
  }

  my $arg = @_ > 0 ? $_[0] : $_;
  
  if(defined $arg && $arg =~ /\.lnk$/ && -r $arg)
  {
    my $target = _win32_resolve(_is_cygwin ? Cygwin::posix_to_win_path($arg) : $arg);
    return $target if defined $target;
  }

  if(_is_cygwin)
  {
    if(@_ > 0)
    { return CORE::readlink($_[0]) }
    else
    # If you need this feature, then please either upgrade to 5.10 or downgrade to 5.6
    { croak "implicit \$_ and Win32::Shortcut::Readlink are not interoperable on Perl 5.8.x" }
  }

  # else is MSWin32
  # emulate unix failues
  if(!defined $arg)
  {
    # TODO: only warn if warnings on in caller
    carp "Use of uninitialized value in readlink";
    $! = 22;
  }
  elsif(-e $arg)
  {
    $! = 22; # Invalid argument
  }
  else
  {
    $! = 2; # No such file or directory
  }
  
  return;
}

1;



lib/Win32/Shortcut/Readlink/Perl510.pm
package
  Win32::Shortcut::Readlink;

use strict;
use warnings;
use Carp qw( carp );
use constant _is_cygwin  => $^O eq 'cygwin';
use constant _is_mswin32 => $^O eq 'MSWin32';
use constant _is_windows => _is_cygwin || _is_mswin32;

sub _real_readlink (_);
# TODO: only warn in 5.14 or earlier if warnings on in caller
*_real_readlink = eval { require 5.016000 } ? \&CORE::readlink : sub (_) { CORE::readlink($_[0]) };

sub readlink (_)
{
  goto &_real_readlink unless _is_windows;
  
  if(defined $_[0] && $_[0] =~ /\.lnk$/ && -r $_[0])
  {
    my $target = _win32_resolve(_is_cygwin ? Cygwin::posix_to_win_path($_[0]) : $_[0]);
    return $target if defined $target;
  }

  goto &_real_readlink if _is_cygwin;

  # else is MSWin32
  # emulate unix failues
  if(!defined $_[0])
  {
    # TODO: only warn if warnings on in caller
    carp "Use of uninitialized value in readlink";
    $! = 22;
  }
  elsif(-e $_[0])
  {
    $! = 22; # Invalid argument
  }
  else
  {
    $! = 2; # No such file or directory
  }
  
  return;
}

1;



lib/Win32/Shortcut/Readlink.pm
package Win32::Shortcut::Readlink;

require 5.008000;
use strict;
use warnings;
use base qw( Exporter );

BEGIN {

# ABSTRACT: Make readlink work with shortcuts
# VERSION

  if($^O =~ /^(cygwin|MSWin32)$/)
  {
    require XSLoader;
    XSLoader::load('Win32::Shortcut::Readlink', $Win32::Shortcut::Readlink::VERSION);
  }

}

=head1 SYNOPSIS

 use Win32::Shortcut::Readlink;
 
 my $target = readlink "c:\\users\\foo\\Desktop\\Application.lnk";

=head1 DESCRIPTION

This module overloads the perl built-in function L<readlink|perlfunc#readlink>
so that it will treat shortcuts like pseudo symlinks on C<cygwin> and C<MSWin32>.
This module doesn't do anything on any other platform, so you are free to make
this a dependency, even if your module or script is going to run on non-Windows
platforms.

This module adjusts the behavior of readlink ONLY in the calling module, so
you shouldn't have to worry about breaking other modules that depend on the
more traditional behavior.

=cut

our @EXPORT_OK = qw( readlink );
our @EXPORT    = @EXPORT_OK;

=head1 FUNCTION

=head2 readlink

 my $target = readlink EXPR
 my $target = readlink

Returns the value of a symbolic link or the target of the shortcut on Windows,
if either symbolic links are implemented or if shortcuts are.  If not, raises an 
exception.  If there is a system error, returns the undefined value and sets 
C<$!> (errno). If C<EXPR> is omitted, uses C<$_>.

=cut

if(eval { require 5.010000 })
{
  require Win32::Shortcut::Readlink::Perl510;
}
elsif(eval { require 5.008000 })
{
  require Win32::Shortcut::Readlink::Perl58;
}
else
{
  # TODO: doesn't currently work.  Figure out
  # if this is a limitation of Perl 5.6 or if
  # I am just doing it wrong.  Fix or remove
  # if appropriate.  BTW- dist requires 5.8
  # so shouldn't even get in here if installed
  # without hacking.
  require Win32::Shortcut::Readlink::Perl56;
}

1;

=head1 CAVEATS

Does not handle Unicode.  Patches welcome.

Before Perl 5.16, C<CORE> functions could not be aliased, and you will see warnings
on Perl 5.14 and earlier if you pass undef in as the argument to readlink, even if
you have warnings turned off.  The work around is to make sure that you never pass
undef to readlink on Perl 5.14 or earlier.

Perl 5.8.x is somewhat supported.  The use of implicit C<$_> with readlink in
Perl 5.8.x is not supported and will throw an exception.  It is recommended that
you either upgrade to at least Perl 5.10 or pass an explicit argument of readlink
when using this module.

=head1 SEE ALSO

=over 4

=item L<Win32::Shortcut>

=item L<Win32::Unicode::Shortcut>

=item L<Win32::Symlink>

=item L<Win32::Hardlink>

=back

=cut
